### ISSUE TITLE
[GSoC 2020] Autoread project weekly reports.

### ISSUE THREAD

**Issue by BK1603** [neovim/neovim#12458]
### WEEK 1

I wrote a plugin that implements fswatch through `vim.loop`:[nvim-autoread](https://github.com/BK1603/nvim-autoread)

This didn't took me that long but since the past few nights I am stuck with this bug that I am facing. What happens is that sometimes, when a file is changed, `checktime` (in the callback) isn't called. And the file is not updated after that specific write.

I tested out everything and believe that either `vim.schedule_wrap()` (maybe the callback is deferred too long and we miss a notification?) or `vim.api.nvim_command`(maybe `checktime` errors out?) is to blame. The fact that we aren't allowed to use `vim.api.nvim_command` inside a lua loop callback makes me a bit more suspicious of it. But as far as I know it could be either of those.

Also why aren't we allowed to call `vim.api.nvim_command` inside lua callbacks? 

@tjdevries (I was advised to tag you for lua related troubles so :p)

**Comment by tjdevries** [comment]
What is the command you want to do w/ `vim.api.nvim_command`?

**Comment by BK1603** [comment]
`checktime`

**Comment by tjdevries** [comment]
You can definitely use `vim.api.nvim_command` in a loop callback if you do something like using `vim.schedule_wrap`, it depends exactly where you mean that you're not allowed to use it. So I don't fully understand exactly what you're saying about that part

You can even use something like `vim.defer_fn` to delay calling until it's "allowed" if you get the notification at a bad time.

This section also worries me:

```lua
  -- sleep for a bit, to ignore multiple notifications from a single change
  -- caused by various editors. (Like (neo)vim :P)
    local timer = uv.new_timer()
    timer:start(1, 0, function()
      timer:stop()
      timer:close()
      WatcherList[fname]:stop()
      WatcherList[fname]:start()
    end)
```

As sleeping for some amount of time seems worrisome. 

I also need more information about exactly the error and scenarios that are coming up that you experience the problem. You could potentially write some tests to reproduce the problem (like, script to open up nvim in one window, and then something edits the file contents using `echo` or similar). That would be helpful for me to give you more targeted information.

**Comment by BK1603** [comment]
> You can definitely use `vim.api.nvim_command` in a loop callback if you do something like using `vim.schedule_wrap`, it depends exactly where you mean that you're not allowed to use it. So I don't fully understand exactly what you're saying about that part

By not allowed I meant that if we use the function without scheduling it, it leads to an error. Why is it so?
> 
> You can even use something like `vim.defer_fn` to delay calling until it's "allowed" if you get the notification at a bad time.

Isn't the function like an alias to uv timer?
> 
> This section also worries me:
> 
> ```lua
>   -- sleep for a bit, to ignore multiple notifications from a single change
>   -- caused by various editors. (Like (neo)vim :P)
>     local timer = uv.new_timer()
>     timer:start(1, 0, function()
>       timer:stop()
>       timer:close()
>       WatcherList[fname]:stop()
>       WatcherList[fname]:start()
>     end)
> ```
> 
> As sleeping for some amount of time seems worrisome.
> 
Well, as I mentioned in the comment we get multiple notifications from a single change on some editors, and this was the only workaround that I could think of for this. :)
Maybe you can help me with this too as I too think that this might make us miss something in a multi user scenario. (Which is where autoread makes the most sense.) :)

> I also need more information about exactly the error and scenarios that are coming up that you experience the problem. You could potentially write some tests to reproduce the problem (like, script to open up nvim in one window, and then something edits the file contents using `echo` or similar). That would be helpful for me to give you more targeted information.

Basically the bug that I am facing is that, the watcher doesn't react to all the changes made to a file. 

I have uploaded the repo with a script for changing the file. It adds 8 lines to the test file. You will see that if we do not use `vim.api.nvim_command` all the changes are printed. (Upto number 8.) But if we do use `vim.api.nvim_command` it is not the case then. 

Another thing that I think I should mention: The if check that I added in the callback, the `listening` one, it only is needed if I call `vim.api.nvim_command`. Somehow when the  command function is called, we try to close the handle twice. Without the command function everything works fine without that check.

**Comment by tjdevries** [comment]
Perhaps you should try something like having one listener for the file that is always around. And what it does is it adds modifications to a queue or some similar structure. Then nvim processes those whenever it has a chance to do so. I am not sure this structure with the constant new file listeners being made is the right structure.

**Comment by BK1603** [comment]
That would definitely help with moving the `checktime` command out of the callback. 

Instead of issuing `checktime` in callback, I will add something similar to an event to a queue. And we can later traverse the queue, just like you said. (I read something about events being sent to multiqueues in `vim.loop`, can we use those?)

But I still don't understand why does this happen? Also, if you don't mind it, can I text you on gitter? I might be sending out too many notifications to others by posting so many comments here. I can later post the approach taken over here in a single comment for everybody else to see.

**Comment by BK1603** [comment]
Well, never mind, using `e!` instead of `checktime` works as expected. I still don't know whats the problem with `checktime`. I will probably look into it later. For now I will continue working on the project. 
Thanks a lot for the help! :)

**Comment by BK1603** [comment]
This week was basically me facing a lot of build issues, but thanks to @teto I was finally able to solve all of them.

The plugin that I made no longer requires wait, I queued the notifications. Thank you @tjdevries for the idea. I used the event handles to add notifications to a queue, and then a check handle for polling the queue. This helped in removing the use of `vim.api.nvim_command` and `vim.schedule_wrap` in the fsevent callback, and with the inconsistency in notifications.

Secondly I added things from https://github.com/neovim/neovim/pull/5599 (TUI events) into the plugin.

Added a few tests for the plugin so that I don't have to run them manually. The plugin is ready now, and I have started working on https://github.com/neovim/neovim/pull/7062 (logging api) so that we can get it merged as it would help immensely with both the plugin and the watcher when it is integrated into the neovim source code.

After the logging PR, I will work on https://github.com/neovim/neovim/pull/5599 so that it can be merged and we can have the intended TUI functionality for autoread in neovim.

I will try and work a bit extra this week if required and get both these PR merged by the next tuesday so that I can start working on integrating the plugin into the source.

**Comment by BK1603** [comment]
# Week 3

The report is a bit late this week since I had to travel for two days and only came back two days ago. Since then I got working on the PRs and almost forgot about this.

## Work done:
I worked on the following PRs in this week:
- #12503 (Should be complete. See the next section for details.)
- #12531 (Writing tests left.)

## Issues faced:
I faced the following issues which took up a lot of time, and weren't exactly solved either:
- **PR #12503**: This took up most of my time. The issue I faced was a test got stuck for seemingly no reason, [Here]() is the line that caused the problems. The conclusion that I reached was that the `spawn` method starts a new session with `is_running = false`. This caused the session to create a blocking request. (From the code [here](https://github.com/neovim/lua-client/blob/master/nvim/session.lua#L81).) I couldn't understand what event was it waiting for. I saw a callback given to `_msg_pack_rpc:write` that calls `uv.stop()`. But I think that happens after the loop is unblocked. (After the previous request is complete.) I tried asking around in the community but couldn't get a lot of help. So in the end I decided to remove the line.

- **TravisCI:** Another issue that I faced in both the PRs was from TravisCI. PR #12503 should be complete, but travis ci can't run the logging tests. At first the error was that the log file wasn't created since the `BUILD_DIR` var was `nil`, which caused `NVIM_LOG_FILE` too to be `nil`. I tried using `TRAVIS_BUILD_DIR` instead of `BUILD_DIR` and now I am getting a `SIGPIPE`.
For #12531, one of the second stage tests seems to fail at random. I tried pushing the same commit twice and two different tests failed.

## Up next:
- Write the remaining tests for #12531
- Start writing tests for file system notification based autoread.

If I am able to do both of these, I will start integrating the plugin to neovim's source. The plugin is ready. What remains is making it adapt to the changes made in #12531 so that there is an option to choose one of those two behaviours. Another problem that I will face down the road is how to determine if filesystem notifications is supported on a particular system or not. (Eg, remote filesystems.) If it isn't fallback to file polling and #12531, else we can rely on the watcher to watch the files and notify for changes.

**Comment by BK1603** [comment]
# Week 5
I apologize for the missing report. I just got too busy during the evaluations and couldn't write it.

## Work Done:
- PR #12531 was merged.
- PR #12503 I was able to resolve the encoding test related error. As I mentioned in the previous report, `module.requiest('nvim_set_client_info', ...)` started a blocking session, I replaced it with `module.nvim_async` and it worked. Though I now have another issue to resolve. [(Build log)](https://travis-ci.org/github/neovim/neovim/jobs/704994839#L2744)
- PR #12593 Started the work on integrating the code. Wrote a simple test for the plugin.

# The upcoming month
The following are my plans for upcoming month

- Complete the integration of the plugin along with the tests. This would consist of the following three things:
    - Integrating the basic plugin as is along with some simple tests. The tests will be for various combinations of the `backup` and `writebackup`  option.
    - Discussing and implementing a fallback mechanism for filesystems that do not support filesystem notifications and writing tests for the said mechanism.
    - Working around platform specific problems, like the absence of the filename parameter (passed to the handle callback) on `OpenBSD`. If no suitable workaround is found, we can fallback to polling for the platform.
- Adding a `fswatch` (name copied from @teto's old work on the same feature) option to neovim, which enables or disables filesystem watchers for reloading files when changed outside the editing session.
- Starting to implement a vimdiff style display for when the user is prompted to reload the file. If possible complete it within this month.

## Upcoming week
For the upcoming week I'd like to integrate the basic plugin along with the tests mentioned above and start working on the fallback tests and mechanism.

**Comment by BK1603** [comment]
# Week 6

## Work Done:
- Added tests for different combinations of `backup` and `writebackup` options. Also added one for `backupcopy`
- Implemented the options for `fswatch`. The option is called `filechangenotify`/`fcnotify` and takes values as follows:
    - **Always:** Always notify.
    - **Changed:** Notify if the buffer contains unsaved changes, else update.
    - **Never:** Never notify, simply update the files.
- The watcher now supports multiple files with the same name.
- The issue regarding `OpenBSD` was solved. Now we don't rely on the arguments to the callback for `filename`. So the watchers should work as expected on all platforms that support file system notifications.
- Added a `showdiff` option to the prompt. When selected it shows the diff between the current buffer and the file on disk.

## Up Next:
- Write tests for the various `fcnotify` options.
- Device a fallback mechanism so that we can fall back to file polling on filesystems that do not support filesystem notifications. This should take a considerable amount of time in my opinion.

**Comment by BK1603** [comment]
# Week 7
Still working on #12593, though I hope to mark it [RDY] soon enough. 

## Issues faced:
- During the tests, `unload` is triggered for the files opened via `edit`. Don't yet know why this happens. I am looking into it right now. It doesn't effect the functionality of a plugin a lot, (it won't at all if I remove the part that tells about trying to close watchers that don't yet exist,) but it shouldn't be happening nonetheless. 
- Failing tests on some the *BSD builds. 
- I have to change the `diff` option so that it opens a new tab and shows the diff there. Though using `:b` reloads the buffer if `autoread` is on, or gives us a prompt similar to what `checktime` gives us when autoread is off.
- #12531 has been causing some troubles with fcnotify. It reloads the file automatically after neovim has gained focus, regardless of whether the user selected `no` at the prompt or not. If I `set noautoread`, then we have two prompts.
- Can't use checktime for reloading the buffers, because it creates a prompt when the current buffer is modified. That leads to us having two different prompts for reloading the same buffer, for a single change.

## Work done:
- Added the missing `off` value to `fcnotify`.
- Made `fcnotify` a list 
- Refactored the code in the PR. (Renamed the files, updated the comments.)

## Up next:
- Fix the issues caused by autoread. As of right now I can't seem to take a clear decision on this one. I could either add another function in lua that calls `buf_reload` but without the prompts. Like a silent checktime. But that would lead to a lot of code repetition. Or edit `buf_check_timestamps` itself. Since it isn't exposed via any API, it shouldn't be a breaking change, but this would mean updating a lot of functions and changing something in the core itself. I will update this comment tomorrow with a better approach if I am able to find one after i wake up and think this through with a fresh mind. 
- Finalize the PR and start fixing failing builds.

## Remote Filesystems
After some discussions it was decided that we should first focus on supporting filesystems that we can support completely, and then later work on other things. So will do exactly that :)

## Notes
Also apparently the first issue that I mentioned with the buffer unloading is because `<afile>` doesn't expand the way I thought it would. When `BufUnload` is triggered for an empty buffer, `<afile>` expands to the name of the buffer that was opened in place of the one that was closed.

**Comment by BK1603** [comment]
# Week 9

Still working on PR https://github.com/neovim/neovim/pull/12593

## Issues faced:
- There was a memory leak happening when the option `fcnotify` was declared as a local option. I couldn't figure out how to fix the leak, but later we decided that the option should be `global-local`, and once I did make it `global-local`, the leak was gone.
- The previous approach of having a separate prompt for the file watcher was causing various inconsistencies with the prompt. We later decided switching the approach a bit so that there is only one `checktime` prompt called in both the cases. This improved a lot of things with `fcnotify`, as described later.

## Work Done:
- The issues caused by autoread were fixed.
- Old build errors and failing tests were fixed.
- The approach to the prompt was redesigned. We now use the `checktime` prompt itself for this instead of creating a new one. This solved a lot of errors with multiple and inconsistent prompts. This also respected what PR https://github.com/neovim/neovim/pull/12531 did, and made sure that https://github.com/neovim/neovim/pull/12531 didn't interfere with the working of `fcnotify`. Also now before notifying the user, we actually check the contents of the file to see if there were any changes.
- Discussed a more sensible approach to integrate the option to neovim. Instead of manipulating `autoread` when `fcnotify` is set, the `buf_check_timestamp` function will be modified so that it respects the value of `fcnotify` along with `autoread`. Also since `changed` is exactly what `autoread` is, but with file system notifications, mention that in the docs.
- Wrote partial docs for the option.

## Up Next:
Since the approach had to be reconsidered due to the various inconsistencies of the previous one, I was set back a bit in terms of my progress. Also some unforeseen incidents happened that costed me a day or two. The PR now stands where it was before we decided switching the approach. What remains again is:
- Fix some failing tests.
- Fix failing builds.
- Finalize the feature and get the PR merged.

I'll try and get this done by the end of next week or earlier.

As for the issue with the `BufUnload` expansion, I've opened a new issue here: https://github.com/neovim/neovim/issues/12690
I don't know if this was the intended behavior, but I don't think it should be as per `:h <afile>` and `:h BufUnload`. If the issue isn't resolved by the time I am done with my project, I will try and take a look at it myself.

**Comment by teto** [comment]
> fcnotify=polling: This will make use of polling for checking for file changes. By itself it will be equivalent to fcnotify=always with file polling.

You don't have to make assumptions: the user could set `set fcnotify=polling,always` to get the behavior you described. Neovim currently defaults autoread to true. Thus I think the default should be `fcnotify=polling,autoread` (for now). Once we confirm filewatching works ok, we can change the default to `fcnotify=watching,autoread` or  `fcnotify=watching,polling.autoread` (polling can be used as a fallback in case the filesystem or platform doesn't support watching)

**Comment by BK1603** [comment]
# Week 10

## Issues faced
- Travis CI reported a memory leak becaues of the newly created option which i was able to fix thanks to @teto 
- Travis logs also report of having a hanged `uv_close_loop`. The problem was supposed to be that the `check` handle and/or the `fs_event` handle was left open which was causing the hang everytime we tried to close the `main_loop`. I tried fixing it by adding a function that closes all the handles at the `VimLeave` autocommand event but that doesn't seem to have fixed things. At first I thought it was because the handles are closed in the next iteration, but then I realized that we do wait for 2 seconds before declaring a hang, so that could not be it. I still don't have a solution for it.
- The hanging `loop_close` has been causing a profiling in the old test suite to fail.

## Work Done
- The implementation of `fcnotify` is now final. It has the values are the same as described in the last report, except that there is an explicit `off` option this time, to have a buffer level control on stopping the watchers and `polling` has now been renamed to `onfocus`.
- `fcnotify` was made a `global-local` option.
- All the failing tests for the option were fixed.
- Tests for the `onfocus` option added.

## Up next
- Fix the `uv_loop_close()` hang
- Make it so that the watcher isn't started at all if the option didn't have `watcher` in it in the first place.

**Comment by BK1603** [comment]
# Week 11

I marked https://github.com/neovim/neovim/pull/12593 ready.
I am now working on https://github.com/neovim/neovim/pull/12503 (the logging PR).

# Work done
- The `uv_loop_close()` hang was fixed.
- `fs_event` watchers are now started only if the option was set to `watcher`.
- Added mechanism for switching backends on the fly. This would help with the addition of more backends for
`fcnotify`. Such as a polling backend on systems that do not support fs events.

# Issues faced:
A timer handle isn't being closed because of `nvim_async` (PR https://github.com/neovim/neovim/pull/12503). [Here](https://travis-ci.org/github/neovim/neovim/jobs/721001528#L7651) are the travis logs. I haven't been able to figure out which handle is causing the problem. It could be one from `session.lua` or from `neovim` itself.

# Up next
- Fix the issue faced with the logging PR.
- Continue working on the `fcnotify` feature. It still can be improved in a lot of ways. For instance, we don't yet have a polling backend which could help on platforms that don't support file system notifications. We are using autocommands an the `FocusGained`/`FocusLost` events for those platforms for now.

**Comment by justinmk** [comment]
# Status

See https://github.com/neovim/neovim/issues/12605
